File: GLLara/GLLLightsListController.m
//
//  GLLLightsListController.m
//  GLLara
//
//  Created by Torsten Kammer on 13.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GLLLightsListController.h"

#import "GLLLightController.h"
#import "LionSubscripting.h"
#import "NSArray+Map.h"

@interface GLLLightsListController ()

@property (nonatomic) NSArray *lights;
- (void)_fetchLights;

@end

@implementation GLLLightsListController

- (id)initWithManagedObjectContext:(NSManagedObjectContext *)managedObjectContext outlineView:(NSOutlineView *)outlineView;
{
    if (!(self = [super init])) return nil;
    
	_managedObjectContext = managedObjectContext;
	_outlineView = outlineView;
		
	return self;
}

- (NSArray *)allSelectableControllers
{
	return self.lights;
}

#pragma mark - Outline view data source

- (id)outlineView:(NSOutlineView *)outlineView child:(NSInteger)index ofItem:(id)item
{
	if (!self.lights) [self _fetchLights];
	return self.lights[index];
}

- (id)outlineView:(NSOutlineView *)outlineView objectValueForTableColumn:(NSTableColumn *)tableColumn byItem:(id)item
{
	if (item == self) return NSLocalizedString(@"Lights", @"source view header");
	
	if ([[item valueForKeyPath:@"entity.name"] isEqual:@"GLLAmbientLight"])
		return NSLocalizedString(@"Ambient", @"source view - lights");
	else
		return [NSString stringWithFormat:NSLocalizedString(@"Diffuse %@", @"source view - lights"), [item valueForKey:@"index"]];
}

- (BOOL)outlineView:(NSOutlineView *)outlineView isItemExpandable:(id)item
{
	return YES;
}

- (NSInteger)outlineView:(NSOutlineView *)outlineView numberOfChildrenOfItem:(id)item
{
	if (!self.lights) [self _fetchLights];
	return self.lights.count;
}

#pragma mark - Outline view delegate

- (BOOL)outlineView:(NSOutlineView *)outlineView isGroupItem:(id)item
{
	return YES;
}

- (BOOL)outlineView:(NSOutlineView *)outlineView shouldSelectItem:(id)item
{
	return NO;
}

#pragma mark - Private methods

- (void)_fetchLights;
{
	NSFetchRequest *lightsRequest = [NSFetchRequest fetchRequestWithEntityName:@"GLLLight"];
	lightsRequest.sortDescriptors = @[ [NSSortDescriptor sortDescriptorWithKey:@"index" ascending:YES] ];
	
	NSArray *lightEntities = [self.managedObjectContext executeFetchRequest:lightsRequest error:NULL];
	
	self.lights = [lightEntities map:^(NSManagedObject *light){
		return [[GLLLightController alloc] initWithLight:light parentController:self];
	}];
}

@end

File: GLLara/GLLLightsListController.swift
//
//  GLLLightsListController.m
//  GLLara
//
//  Created by Torsten Kammer on 13.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

/**
 * Source list controller for a list of lights, and direct child of the root.
 */
@objc class GLLLightsListController: NSObject, NSOutlineViewDataSource, NSOutlineViewDelegate {
    @objc init(managedObjectContext: NSManagedObjectContext, outlineView: NSOutlineView) {
        self.managedObjectContext = managedObjectContext
        self.outlineView = outlineView
        
        super.init()
        
        let request = NSFetchRequest<NSManagedObject>(entityName: "GLLLight")
        request.sortDescriptors = [ NSSortDescriptor(key: "index", ascending: true) ]
        
        lights = try! managedObjectContext.fetch(request).map { GLLLightController(light: $0, parentController: self) }
    }
    
    @objc let managedObjectContext: NSManagedObjectContext
    @objc let outlineView: NSOutlineView
    
    var lights: [GLLLightController] = []
    
    @objc var allSelectableControllers: [GLLLightController] {
        return lights
    }
    
    func outlineView(_ outlineView: NSOutlineView, child index: Int, ofItem item: Any?) -> Any {
        return lights[index]
    }
    
    func outlineView(_ outlineView: NSOutlineView, objectValueFor tableColumn: NSTableColumn?, byItem item: Any?) -> Any? {
        return NSLocalizedString("Lights", comment: "source view header")
    }
    
    func outlineView(_ outlineView: NSOutlineView, isItemExpandable item: Any) -> Bool {
        return true
    }
    
    func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
        return lights.count
    }
    
    func outlineView(_ outlineView: NSOutlineView, isGroupItem item: Any) -> Bool {
        return true
    }
    
    func outlineView(_ outlineView: NSOutlineView, shouldSelectItem item: Any) -> Bool {
        return false
    }
}
